File: test/extensions/language/choice_types/python/UInt64ParamChoiceTest.py
import os
import zserio

import ChoiceTypes

from testutils import getApiDir


class UInt64ParamChoiceTest(ChoiceTypes.TestCase):
    def testConstructor(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(self.VARIANT_A_SELECTOR, uint64ParamChoice.selector)

        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR, value_b_=1234)
        self.assertEqual(self.VARIANT_B_SELECTOR, uint64ParamChoice.selector)
        self.assertEqual(1234, uint64ParamChoice.value_b)

    def testFromReader(self):
        selector = self.VARIANT_B_SELECTOR
        value = 234
        writer = zserio.BitStreamWriter()
        UInt64ParamChoiceTest._writeUInt64ParamChoiceToStream(writer, selector, value)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        uint64ParamChoice = self.api.UInt64ParamChoice.from_reader(reader, selector)
        self.assertEqual(selector, uint64ParamChoice.selector)
        self.assertEqual(value, uint64ParamChoice.value_b)

    def testEq(self):
        uint64ParamChoice1 = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        uint64ParamChoice2 = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertTrue(uint64ParamChoice1 == uint64ParamChoice2)

        value = 99
        uint64ParamChoice1.value_a = value
        self.assertFalse(uint64ParamChoice1 == uint64ParamChoice2)

        uint64ParamChoice2.value_a = value
        self.assertTrue(uint64ParamChoice1 == uint64ParamChoice2)

        diffValue = value + 1
        uint64ParamChoice2.value_a = diffValue
        self.assertFalse(uint64ParamChoice1 == uint64ParamChoice2)

    def testHash(self):
        uint64ParamChoice1 = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        uint64ParamChoice2 = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(hash(uint64ParamChoice1), hash(uint64ParamChoice2))

        value = 99
        uint64ParamChoice1.value_a = value
        self.assertTrue(hash(uint64ParamChoice1) != hash(uint64ParamChoice2))

        uint64ParamChoice2.value_a = value
        self.assertEqual(hash(uint64ParamChoice1), hash(uint64ParamChoice2))

        diffValue = value + 1
        uint64ParamChoice2.value_a = diffValue
        self.assertTrue(hash(uint64ParamChoice1) != hash(uint64ParamChoice2))

        # use hardcoded values to check that the hash code is stable
        # using __hash__ to prevent 32-bit Python hash() truncation
        self.assertEqual(31623, uint64ParamChoice1.__hash__())
        self.assertEqual(31624, uint64ParamChoice2.__hash__())

    def testGetSelector(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_C_SELECTOR)
        self.assertEqual(self.VARIANT_C_SELECTOR, uint64ParamChoice.selector)

    def testGetSetA(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        value = 99
        uint64ParamChoice.value_a = value
        self.assertEqual(value, uint64ParamChoice.value_a)

    def testGetSetB(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR)
        value = 234
        uint64ParamChoice.value_b = value
        self.assertEqual(value, uint64ParamChoice.value_b)

    def testGetSetC(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_C_SELECTOR)
        value = 23456
        uint64ParamChoice.value_c = value
        self.assertEqual(value, uint64ParamChoice.value_c)

    def testChoiceTag(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(uint64ParamChoice.CHOICE_VALUE_A, uint64ParamChoice.choice_tag)

        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR)
        self.assertEqual(uint64ParamChoice.CHOICE_VALUE_B, uint64ParamChoice.choice_tag)

        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_C_SELECTOR)
        self.assertEqual(uint64ParamChoice.CHOICE_VALUE_C, uint64ParamChoice.choice_tag)

        uint64ParamChoice = self.api.UInt64ParamChoice(self.EMPTY_SELECTOR)
        self.assertEqual(uint64ParamChoice.UNDEFINED_CHOICE, uint64ParamChoice.choice_tag)

    def testBitSizeOf(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        self.assertEqual(8, uint64ParamChoice.bitsizeof())

        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR)
        self.assertEqual(16, uint64ParamChoice.bitsizeof())

    def testInitializeOffsets(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        bitPosition = 1
        self.assertEqual(9, uint64ParamChoice.initialize_offsets(bitPosition))

        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR)
        self.assertEqual(17, uint64ParamChoice.initialize_offsets(bitPosition))

    def testWriteRead(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        byteValue = 99
        uint64ParamChoice.value_a = byteValue
        writer = zserio.BitStreamWriter()
        uint64ParamChoice.write(writer)
        readUInt64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readUInt64ParamChoice.read(reader)
        self.assertEqual(byteValue, readUInt64ParamChoice.value_a)
        self.assertEqual(uint64ParamChoice, readUInt64ParamChoice)

        shortValue = 234
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR, value_b_=shortValue)
        writer = zserio.BitStreamWriter()
        uint64ParamChoice.write(writer)
        readUInt64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR)
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readUInt64ParamChoice.read(reader)
        self.assertEqual(shortValue, readUInt64ParamChoice.value_b)
        self.assertEqual(uint64ParamChoice, readUInt64ParamChoice)

    def testWriteReadFile(self):
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_A_SELECTOR)
        byteValue = 99
        uint64ParamChoice.value_a = byteValue
        filenameA = self.BLOB_NAME_BASE + "a.blob"
        zserio.serialize_to_file(uint64ParamChoice, filenameA)

        readUInt64ParamChoice = zserio.deserialize_from_file(
            self.api.UInt64ParamChoice, filenameA, self.VARIANT_A_SELECTOR
        )
        self.assertEqual(byteValue, readUInt64ParamChoice.value_a)
        self.assertEqual(uint64ParamChoice, readUInt64ParamChoice)

        shortValue = 234
        uint64ParamChoice = self.api.UInt64ParamChoice(self.VARIANT_B_SELECTOR, value_b_=shortValue)
        filenameB = self.BLOB_NAME_BASE + "b.blob"
        zserio.serialize_to_file(uint64ParamChoice, filenameB)

        readUInt64ParamChoice = zserio.deserialize_from_file(
            self.api.UInt64ParamChoice, filenameB, self.VARIANT_B_SELECTOR
        )
        self.assertEqual(shortValue, readUInt64ParamChoice.value_b)
        self.assertEqual(uint64ParamChoice, readUInt64ParamChoice)

    @staticmethod
    def _writeUInt64ParamChoiceToStream(writer, selector, value):
        if selector == 1:
            writer.write_signed_bits(value, 8)
        elif selector in (2, 3, 4):
            writer.write_signed_bits(value, 16)
        elif selector in (5, 6):
            pass
        else:
            writer.write_signed_bits(value, 32)

    BLOB_NAME_BASE = os.path.join(getApiDir(os.path.dirname(__file__)), "uint64_param_choice_")
    VARIANT_A_SELECTOR = 1
    VARIANT_B_SELECTOR = 2
    VARIANT_C_SELECTOR = 7
    EMPTY_SELECTOR = 5

File: test/extensions/language/choice_types/java/choice_types/UInt64ParamChoiceTest.java
package choice_types;

import static org.junit.jupiter.api.Assertions.*;

import java.io.File;
import java.io.IOException;
import java.math.BigInteger;

import org.junit.jupiter.api.Test;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.SerializeUtil;

import choice_types.uint64_param_choice.UInt64ParamChoice;

public class UInt64ParamChoiceTest
{
    @Test
    public void selectorConstructor()
    {
        final UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        assertEquals(VARIANT_A_SELECTOR, uint64ParamChoice.getSelector());
    }

    @Test
    public void bitStreamReaderConstructor() throws IOException, ZserioError
    {
        final short value = 234;
        final BitBuffer buffer = writeUInt64ParamChoiceToBitBuffer(VARIANT_B_SELECTOR, value);
        final ByteArrayBitStreamReader stream =
                new ByteArrayBitStreamReader(buffer.getBuffer(), buffer.getBitSize());
        final UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(stream, VARIANT_B_SELECTOR);
        stream.close();
        assertEquals(VARIANT_B_SELECTOR, uint64ParamChoice.getSelector());
        assertEquals(value, uint64ParamChoice.getValueB());
    }

    @Test
    public void choiceTag()
    {
        UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        assertEquals(UInt64ParamChoice.CHOICE_valueA, uint64ParamChoice.choiceTag());

        uint64ParamChoice = new UInt64ParamChoice(VARIANT_B_SELECTOR);
        assertEquals(UInt64ParamChoice.CHOICE_valueB, uint64ParamChoice.choiceTag());

        uint64ParamChoice = new UInt64ParamChoice(VARIANT_C_SELECTOR);
        assertEquals(UInt64ParamChoice.CHOICE_valueC, uint64ParamChoice.choiceTag());

        uint64ParamChoice = new UInt64ParamChoice(EMPTY_SELECTOR);
        assertEquals(UInt64ParamChoice.UNDEFINED_CHOICE, uint64ParamChoice.choiceTag());
    }

    @Test
    public void bitSizeOf()
    {
        UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        assertEquals(8, uint64ParamChoice.bitSizeOf());

        uint64ParamChoice = new UInt64ParamChoice(VARIANT_B_SELECTOR);
        assertEquals(16, uint64ParamChoice.bitSizeOf());
    }

    @Test
    public void getSelector()
    {
        final UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(VARIANT_C_SELECTOR);
        assertEquals(VARIANT_C_SELECTOR, uint64ParamChoice.getSelector());
    }

    @Test
    public void getSetA()
    {
        UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        final byte value = 99;
        uint64ParamChoice.setValueA(value);
        assertEquals(value, uint64ParamChoice.getValueA());
    }

    @Test
    public void getSetB()
    {
        UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(VARIANT_B_SELECTOR);
        final short value = 234;
        uint64ParamChoice.setValueB(value);
        assertEquals(value, uint64ParamChoice.getValueB());
    }

    @Test
    public void getSetC()
    {
        UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(VARIANT_C_SELECTOR);
        final int value = 23456;
        uint64ParamChoice.setValueC(value);
        assertEquals(value, uint64ParamChoice.getValueC());
    }

    @Test
    public void equals()
    {
        UInt64ParamChoice uint64ParamChoice1 = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        UInt64ParamChoice uint64ParamChoice2 = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        assertTrue(uint64ParamChoice1.equals(uint64ParamChoice2));

        final byte value = 99;
        uint64ParamChoice1.setValueA(value);
        assertFalse(uint64ParamChoice1.equals(uint64ParamChoice2));

        uint64ParamChoice2.setValueA(value);
        assertTrue(uint64ParamChoice1.equals(uint64ParamChoice2));

        final byte diffValue = value + 1;
        uint64ParamChoice2.setValueA(diffValue);
        assertFalse(uint64ParamChoice1.equals(uint64ParamChoice2));
    }

    @Test
    public void hashCodeMethod()
    {
        UInt64ParamChoice uint64ParamChoice1 = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        UInt64ParamChoice uint64ParamChoice2 = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        assertEquals(uint64ParamChoice1.hashCode(), uint64ParamChoice2.hashCode());

        final byte value = 99;
        uint64ParamChoice1.setValueA(value);
        assertTrue(uint64ParamChoice1.hashCode() != uint64ParamChoice2.hashCode());

        uint64ParamChoice2.setValueA(value);
        assertEquals(uint64ParamChoice1.hashCode(), uint64ParamChoice2.hashCode());

        final byte diffValue = value + 1;
        uint64ParamChoice2.setValueA(diffValue);
        assertTrue(uint64ParamChoice1.hashCode() != uint64ParamChoice2.hashCode());

        // use hardcoded values to check that the hash code is stable
        assertEquals(31623, uint64ParamChoice1.hashCode());
        assertEquals(31624, uint64ParamChoice2.hashCode());
    }

    @Test
    public void initializeOffsets()
    {
        UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        final int bitPosition = 1;
        assertEquals(9, uint64ParamChoice.initializeOffsets(bitPosition));

        uint64ParamChoice = new UInt64ParamChoice(VARIANT_B_SELECTOR);
        assertEquals(17, uint64ParamChoice.initializeOffsets(bitPosition));
    }

    @Test
    public void writeRead() throws IOException, ZserioError
    {
        UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        final byte byteValue = 99;
        uint64ParamChoice.setValueA(byteValue);
        ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        uint64ParamChoice.write(writer);
        ByteArrayBitStreamReader reader =
                new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        UInt64ParamChoice readUInt64ParamChoice = new UInt64ParamChoice(reader, VARIANT_A_SELECTOR);
        assertEquals(byteValue, readUInt64ParamChoice.getValueA());

        uint64ParamChoice = new UInt64ParamChoice(VARIANT_B_SELECTOR);
        final short shortValue = 234;
        uint64ParamChoice.setValueB(shortValue);
        writer = new ByteArrayBitStreamWriter();
        uint64ParamChoice.write(writer);
        reader = new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        readUInt64ParamChoice = new UInt64ParamChoice(reader, VARIANT_B_SELECTOR);
        assertEquals(shortValue, readUInt64ParamChoice.getValueB());
    }

    @Test
    public void writeReadFile() throws IOException, ZserioError
    {
        UInt64ParamChoice uint64ParamChoice = new UInt64ParamChoice(VARIANT_A_SELECTOR);
        final byte byteValue = 99;
        uint64ParamChoice.setValueA(byteValue);
        final File fileA = new File(BLOB_NAME_BASE + "a.blob");
        SerializeUtil.serializeToFile(uint64ParamChoice, fileA);
        UInt64ParamChoice readUInt64ParamChoice =
                SerializeUtil.deserializeFromFile(UInt64ParamChoice.class, fileA, VARIANT_A_SELECTOR);
        assertEquals(byteValue, readUInt64ParamChoice.getValueA());

        uint64ParamChoice = new UInt64ParamChoice(VARIANT_B_SELECTOR);
        final short shortValue = 234;
        uint64ParamChoice.setValueB(shortValue);
        final File fileB = new File(BLOB_NAME_BASE + "b.blob");
        SerializeUtil.serializeToFile(uint64ParamChoice, fileB);
        readUInt64ParamChoice =
                SerializeUtil.deserializeFromFile(UInt64ParamChoice.class, fileB, VARIANT_B_SELECTOR);
        assertEquals(shortValue, readUInt64ParamChoice.getValueB());
    }

    private BitBuffer writeUInt64ParamChoiceToBitBuffer(BigInteger selector, int value) throws IOException
    {
        try (final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            if (selector.compareTo(new BigInteger("1")) == 0)
                writer.writeByte((byte)value);
            else if (selector.compareTo(new BigInteger("2")) == 0 ||
                    selector.compareTo(new BigInteger("3")) == 0 ||
                    selector.compareTo(new BigInteger("4")) == 0)
                writer.writeShort((short)value);
            else if (selector.compareTo(new BigInteger("5")) == 0 ||
                    selector.compareTo(new BigInteger("6")) == 0)
                ;
            else
                writer.writeInt(value);

            return new BitBuffer(writer.toByteArray(), writer.getBitPosition());
        }
    }

    private static final String BLOB_NAME_BASE = "uint64_param_choice_";
    private static final BigInteger VARIANT_A_SELECTOR = BigInteger.ONE;
    private static final BigInteger VARIANT_B_SELECTOR = new BigInteger("2");
    private static final BigInteger VARIANT_C_SELECTOR = new BigInteger("7");
    private static final BigInteger EMPTY_SELECTOR = new BigInteger("5");
}
